test/integration/node/lifecycle_test.go

```diff
@@ -44,6 +49,290 @@ const poll = 2 * time.Second
+			}
+			testPod := &v1.Pod{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "testpod",
+				},
+				Spec: v1.PodSpec{
+					NodeName: nodes[nodeIndex].Name,
+					Containers: []v1.Container{
+						{Name: "container", Image: imageutils.GetPauseImageName()},
+					},
+				},
+				Status: v1.PodStatus{
+					Phase: v1.PodRunning,
+					Conditions: []v1.PodCondition{
+						{
+							Type:   v1.PodReady,
+							Status: v1.ConditionTrue,
+						},
+					},
+				},
+			}
+
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.JobBackoffPolicy, test.enableJobBackoffPolicy)()
+			testCtx := testutils.InitTestAPIServer(t, "taint-no-execute", nil)
+
+			// Build clientset and informers for controllers.
+			defer testutils.CleanupTest(t, testCtx)
+			cs := testCtx.ClientSet
+
+			// Build clientset and informers for controllers.
+			externalClientConfig := restclient.CopyConfig(testCtx.KubeConfig)
+			externalClientConfig.QPS = -1
+			externalClientset := clientset.NewForConfigOrDie(externalClientConfig)
+			externalInformers := informers.NewSharedInformerFactory(externalClientset, time.Second)
+
+			// Start NodeLifecycleController for taint.
+			nc, err := nodelifecycle.NewNodeLifecycleController(
+				testCtx.Ctx,
+				externalInformers.Coordination().V1().Leases(),
+				externalInformers.Core().V1().Pods(),
+				externalInformers.Core().V1().Nodes(),
+				externalInformers.Apps().V1().DaemonSets(),
+				cs,
+				1*time.Second,    // Node monitor grace period
+				time.Minute,      // Node startup grace period
+				time.Millisecond, // Node monitor period
+				1,                // Pod eviction timeout
+				100,              // Eviction limiter QPS
+				100,              // Secondary eviction limiter QPS
+				50,               // Large cluster threshold
+				0.55,             // Unhealthy zone threshold
+				true,             // Run taint manager
+			)
+			if err != nil {
+				t.Errorf("Failed to create node controller: %v", err)
+				return
+			}
+
+			// Waiting for all controllers to sync
+			externalInformers.Start(testCtx.Ctx.Done())
+			externalInformers.WaitForCacheSync(testCtx.Ctx.Done())
+
+			// Run all controllers
+			go nc.Run(testCtx.Ctx)
+
+			for index := range nodes {
+				nodes[index], err = cs.CoreV1().Nodes().Create(testCtx.Ctx, nodes[index], metav1.CreateOptions{})
+				if err != nil {
+					t.Errorf("Failed to create node, err: %v", err)
+				}
+			}
+
+			testPod, err = cs.CoreV1().Pods(testCtx.NS.Name).Create(testCtx.Ctx, testPod, metav1.CreateOptions{})
+			if err != nil {
+				t.Fatalf("Test Failed: error: %v, while creating pod", err)
+			}
+
+			if err := testutils.AddTaintToNode(cs, nodes[nodeIndex].Name, v1.Taint{Key: "AnyNode", Effect: v1.TaintEffectNoExecute}); err != nil {
+				t.Errorf("Failed to taint node in test %s <%s>, err: %v", name, nodes[nodeIndex].Name, err)
+			}
+
+			err = wait.PollImmediate(time.Second, time.Second*20, func() (bool, error) {
```

Comment:
The wait should be for the pod to have a deletion timestamp.

After that, outside of the wait, you can check whether the pod has or doesn't have the condition.

Otherwise, your check for `!test.enableJobBackoffPolicy && cond == nil` would be true even if the controller does nothing.

Reply:
indeed, fixed